src/Libraries/Microsoft.Extensions.Telemetry.Testing/Metering/MetricCollector.cs

```diff
@@ -200,57 +202,58 @@ public IReadOnlyList<CollectedMeasurement<T>> GetMeasurementSnapshot(bool clear
-    public Task WaitForMeasurementsAsync(int numMeasurements)
+    public Task WaitForMeasurementsAsync(int minCount, CancellationToken cancellationToken = default)
     {
-        _ = Throw.IfLessThan(numMeasurements, 1);
+        _ = Throw.IfLessThan(minCount, 1);
+
+        ThrowIfDisposed();
 
+        Waiter w;
         lock (_measurements)
         {
-            ThrowIfDisposed();
-
-            if (_measurements.Count >= numMeasurements)
+            if (_measurements.Count >= minCount)
             {
                 return Task.CompletedTask;
             }
 
-            var w = new Waiter(numMeasurements);
+            w = new Waiter(minCount);
             _waiters.Add(w);
+        }
 
-            return w.TaskSource.Task;
+        if (cancellationToken.CanBeCanceled)
+        {
+            _ = cancellationToken.Register(() =>
+            {
+                lock (_measurements)
+                {
+#if NET6_0_OR_GREATER
+                    w.TaskSource.TrySetCanceled(cancellationToken);
+#else
+                    w.TaskSource.TrySetCanceled();
+#endif
+
+                    _waiters.Remove(w);
+                }
+            });
         }
+
+        return w.TaskSource.Task;
     }
 
     /// <summary>
     /// Returns a task that completes when the collector has collected a minimum number of measurements.
     /// </summary>
-    /// <param name="numMeasurements">The minimum number of measurements to wait for.</param>
+    /// <param name="minCount">The minimum number of measurements to wait for.</param>
     /// <param name="timeout">How long to wait.</param>
-    /// <param name="timeProvider">The time provider against which the timeout executes. If this is <see langword="null"/> then the provider
-    /// selected when creating the metric collector is used.</param>
     /// <returns>A task that completes when the collector has collected the requisite number of measurements.</returns>
     [SuppressMessage("Resilience", "R9A061:The async method doesn't support cancellation", Justification = "Not relevant in this case")]
-    public Task WaitForMeasurementsAsync(int numMeasurements, TimeSpan timeout, TimeProvider? timeProvider = null)
+    public async Task WaitForMeasurementsAsync(int minCount, TimeSpan timeout)
     {
-        _ = Throw.IfLessThan(numMeasurements, 1);
-        _ = Throw.IfLessThan(timeout.Ticks, 0);
-
-        lock (_measurements)
-        {
-            ThrowIfDisposed();
-
-            if (_measurements.Count >= numMeasurements)
-            {
-                return Task.CompletedTask;
-            }
-
-            var w = new Waiter(numMeasurements);
-            _waiters.Add(w);
-
-            return w.TaskSource.Task.WaitAsync(timeout, timeProvider ?? _timeProvider, CancellationToken.None);
-        }
+        using var cancellationTokenSource = new CancellationTokenSource(timeout);
+        await WaitForMeasurementsAsync(minCount, cancellationTokenSource.Token).ConfigureAwait(false);
```

Comment:
Why the await here? Can't this just be tail-optimized to just return a task without the await?

Reply:
Because the CTS is disposed with the using:

```csharp
using var cancellationTokenSource = new CancellationTokenSource(timeout);
```

Without the await then the CTS would immediately be disposed when the method exits.